Alarm raised for aggregate lambda errors during a reindex

CloudWatch logs on anvildev:
- [RequestId 9fda45c5](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2023-08-31T19*3a24*3a13.988Z~start~'2023-08-31T19*3a22*3a50.310Z~timeType~'ABSOLUTE~tz~'UTC~editorString~'fields*20*40timestamp*2c*20*40message*0a*7c*20filter*20*40requestId*20*3d*20*279fda45c5-e898-5ae8-be30-18af025a26aa*27*0a*7c*20sort*20*40timestamp*20asc*0a*7c*20limit*20200~queryId~'e9cc717fbf39df8-4aca2dde-4e5c55d-39baaf36-70e9f21c85c0af7ca45ae5d~source~(~'*2faws*2flambda*2fazul-indexer-anvildev-aggregate)))

```
[
    {
        "@timestamp": "2023-08-31 19:23:02.158",
        "@message": "START RequestId: 9fda45c5-e898-5ae8-be30-18af025a26aa Version: $LATEST\n"
    },
    {
        "@timestamp": "2023-08-31 19:23:02.159",
        "@message": "[INFO]\t2023-08-31T19:23:02.159Z\t9fda45c5-e898-5ae8-be30-18af025a26aa\tazul.indexer.index_controller\tAttempt 1 of handling 1 contribution(s) for entity anvil/datasets/677dd55c-3fa3-4b07-8c98-985d94d7577e"
    },
    {
        "@timestamp": "2023-08-31 19:23:02.159",
        "@message": "[INFO]\t2023-08-31T19:23:02.159Z\t9fda45c5-e898-5ae8-be30-18af025a26aa\tazul.indexer.index_controller\tAttempt 1 of handling 1 contribution(s) for entity anvil/donors/f1da22e5-fca8-4fd6-9abb-f8280713bc83"
    },
    {
        "@timestamp": "2023-08-31 19:23:02.159",
        "@message": "[INFO]\t2023-08-31T19:23:02.159Z\t9fda45c5-e898-5ae8-be30-18af025a26aa\tazul.indexer.index_controller\tAttempt 1 of handling 1 contribution(s) for entity anvil/files/241322f6-a3ff-46ae-a499-cc41b9ae7a34"
    },
…
    {
        "@timestamp": "2023-08-31 19:23:57.563",
        "@message": "[INFO]\t2023-08-31T19:23:57.562Z\t9fda45c5-e898-5ae8-be30-18af025a26aa\tazul.indexer.index_service\tRead 6350 contribution(s)"
    },
    {
        "@timestamp": "2023-08-31 19:23:57.883",
        "@message": "[DEBUG]\t2023-08-31T19:23:57.883Z\t9fda45c5-e898-5ae8-be30-18af025a26aa\tazul.indexer.index_service\tNumber of contributions read, by entity: {'datasets/677dd55c-3fa3-4b07-8c98-985d94d7577e': 6336, 'donors/f1da22e5-fca8-4fd6-9abb-f8280713bc83': 2, 'files/241322f6-a3ff-46ae-a499-cc41b9ae7a34': 2, 'files/26bbb642-fb60-4ea2-86bc-832c8b98efce': 2, 'files/282eecd9-6ae5-444b-8001-ac5ac7760092': 1, 'files/67fb6f6d-bd84-4c54-9271-273297d862e1': 2, 'files/aa4ec0db-2219-4cf0-bc39-f5696104bd26': 1, 'files/d31c0b94-c326-47ab-b3b7-3a119eec5ceb': 1, 'files/d5805d9d-e338-4cdf-a9fc-d6c444da784b': 1, 'files/efcc84a6-8e9a-454b-a809-149ae26c166e': 2}"
    },
    {
        "@timestamp": "2023-08-31 19:23:58.521",
        "@message": "[INFO]\t2023-08-31T19:23:58.520Z\t9fda45c5-e898-5ae8-be30-18af025a26aa\tazul.indexer.index_service\tSelected 6350 contribution(s) to be aggregated."
    },
    {
        "@timestamp": "2023-08-31 19:23:58.521",
        "@message": "[DEBUG]\t2023-08-31T19:23:58.521Z\t9fda45c5-e898-5ae8-be30-18af025a26aa\tazul.indexer.index_service\tNumber of contributions selected for aggregation, by entity: {'datasets/677dd55c-3fa3-4b07-8c98-985d94d7577e': 6336, 'donors/f1da22e5-fca8-4fd6-9abb-f8280713bc83': 2, 'files/241322f6-a3ff-46ae-a499-cc41b9ae7a34': 2, 'files/26bbb642-fb60-4ea2-86bc-832c8b98efce': 2, 'files/282eecd9-6ae5-444b-8001-ac5ac7760092': 1, 'files/67fb6f6d-bd84-4c54-9271-273297d862e1': 2, 'files/aa4ec0db-2219-4cf0-bc39-f5696104bd26': 1, 'files/d31c0b94-c326-47ab-b3b7-3a119eec5ceb': 1, 'files/d5805d9d-e338-4cdf-a9fc-d6c444da784b': 1, 'files/efcc84a6-8e9a-454b-a809-149ae26c166e': 2}"
    },
    {
        "@timestamp": "2023-08-31 19:24:02.282",
        "@message": "2023-08-31T19:24:02.282Z 9fda45c5-e898-5ae8-be30-18af025a26aa Task timed out after 60.12 seconds\n\n"
    },
    {
        "@timestamp": "2023-08-31 19:24:02.282",
        "@message": "END RequestId: 9fda45c5-e898-5ae8-be30-18af025a26aa\n"
    },
    {
        "@timestamp": "2023-08-31 19:24:02.282",
        "@message": "REPORT RequestId: 9fda45c5-e898-5ae8-be30-18af025a26aa\tDuration: 60122.92 ms\tBilled Duration: 60000 ms\tMemory Size: 256 MB\tMax Memory Used: 256 MB\t\n"
    }
]
```

Other occurrences during the same reindex:
- RequestId [b66b0138](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2023-08-31T19*3a25*3a19.764Z~start~'2023-08-31T19*3a24*3a01.409Z~timeType~'ABSOLUTE~tz~'UTC~editorString~'fields*20*40timestamp*2c*20*40message*0a*7c*20filter*20*40requestId*20*3d*20*27b66b0138-b9cc-5d56-8deb-5c5252c77f0f*27*0a*7c*20sort*20*40timestamp*20asc*0a*7c*20limit*20200~queryId~'e9cc717fbf39df8-4aca2dde-4e5c55d-39baaf36-70e9f21c85c0af7ca45ae5d~source~(~'*2faws*2flambda*2fazul-indexer-anvildev-aggregate)))
- RequestId [a0118205](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2023-08-31T19*3a27*3a14.741Z~start~'2023-08-31T19*3a24*3a38.706Z~timeType~'ABSOLUTE~tz~'UTC~editorString~'fields*20*40timestamp*2c*20*40message*0a*7c*20filter*20*40requestId*20*3d*20*27a0118205-8752-53a0-84ca-df6f6c036651*27*0a*7c*20sort*20*40timestamp*20asc*0a*7c*20limit*20200~queryId~'e9cc717fbf39df8-4aca2dde-4e5c55d-39baaf36-70e9f21c85c0af7ca45ae5d~source~(~'*2faws*2flambda*2fazul-indexer-anvildev-aggregate)))
- RequestId [cba833ea](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2023-08-31T19*3a27*3a50.887Z~start~'2023-08-31T19*3a26*3a22.931Z~timeType~'ABSOLUTE~tz~'UTC~editorString~'fields*20*40timestamp*2c*20*40message*0a*7c*20filter*20*40requestId*20*3d*20*27cba833ea-61ad-5231-88dc-6e808a37f251*27*0a*7c*20sort*20*40timestamp*20asc*0a*7c*20limit*20200~queryId~'e9cc717fbf39df8-4aca2dde-4e5c55d-39baaf36-70e9f21c85c0af7ca45ae5d~source~(~'*2faws*2flambda*2fazul-indexer-anvildev-aggregate)))
- RequestId [0338f0a1](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2023-08-31T19*3a29*3a12.224Z~start~'2023-08-31T19*3a27*3a33.793Z~timeType~'ABSOLUTE~tz~'UTC~editorString~'fields*20*40timestamp*2c*20*40message*0a*7c*20filter*20*40requestId*20*3d*20*270338f0a1-59d9-55c5-b74c-adcbb23b254c*27*0a*7c*20sort*20*40timestamp*20asc*0a*7c*20limit*20200~queryId~'e9cc717fbf39df8-4aca2dde-4e5c55d-39baaf36-70e9f21c85c0af7ca45ae5d~source~(~'*2faws*2flambda*2fazul-indexer-anvildev-aggregate)))

-----

- [ ] Security design review completed; the Resolution of this issue does **not** …
  - [ ] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [ ] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [ ] … affect the permissions of external users like access to
    - TDR snapshots
  - [ ] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [ ] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [ ] … affect monitoring of the system
  - [ ] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [ ] … add an interface that exposes sensitive or confidential data at the security boundary
  - [ ] … affect the encryption of data at rest
  - [ ] … require persistence of sensitive or confidential data that might require encryption at rest
  - [ ] … require unencrypted transmission of data within the security boundary
  - [ ] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [ ] Documentation on any unchecked boxes is provided in comments below

> The non-retry lambda is expected to time out. Assignee to clarify

Investigation discovered a bug in the configuration of the Lambda alarms, all are being created with a threshold of zero.

Generating terraform configs for `anvildev`:
```
.venv) daniel@Crispin ~/repo/azul3 $ _select
lrwxr-xr-x  1 daniel  staff  8 Sep  7 09:40 .active -> anvildev
(.venv) daniel@Crispin ~/repo/azul3 $
(.venv) daniel@Crispin ~/repo/azul3 $ make deploy
…
Plan: 1 to add, 11 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: n

Apply cancelled.
make[1]: *** [apply] Error 1
make: *** [terraform] Error 2
```

Checking thresholds set for each alarm, note all "_errors" and "_throttles" alarms have a threshold of `0`:
```
(.venv) daniel@Crispin ~/repo/azul3 $ egrep 'alarm_name|threshold' terraform/cloudwatch.tf.json
                        "alarm_name": "azul-indexer_5xx-anvildev",
                        "threshold": 1,
                        "alarm_name": "azul-service_5xx-anvildev",
                        "threshold": 1,
                        "alarm_name": "azul-indexercachehealth-anvildev.alarm",
                        "threshold": 1,
                        "alarm_name": "azul-servicecachehealth-anvildev.alarm",
                        "threshold": 1,
                        "alarm_name": "azul-internet_ingress-anvildev",
                        "threshold": 53687091200,
                        "alarm_name": "azul-internet_egress-anvildev",
                        "threshold": 10737418240,
                        "alarm_name": "azul-vpn_ingress-anvildev",
                        "threshold": 107374182400,
                        "alarm_name": "azul-vpn_egress-anvildev",
                        "threshold": 10737418240,
                        "alarm_name": "azul-indexer_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_indexercachehealth_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_contribute_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_aggregate_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_aggregate_retry_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_contribute_retry_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_forward_alb_logs_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_forward_s3_logs_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_indexercachehealth_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_contribute_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_aggregate_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_aggregate_retry_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_contribute_retry_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_forward_alb_logs_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-indexer_forward_s3_logs_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-service_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-service_servicecachehealth_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-service_manifest_errors-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-service_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-service_servicecachehealth_throttles-anvildev.alarm",
                        "threshold": 0,
                        "alarm_name": "azul-service_manifest_throttles-anvildev.alarm",
                        "threshold": 0,
```

Further investigation found the cause to be the placement of the `@app.threshold` decorator.

Swapping the current order of decorators:
```
@app.on_sqs_message(queue=config.tallies_queue_name(),
                    batch_size=IndexController.document_batch_size)
@app.threshold(errors=int(config.aggregation_concurrency(retry=False) * 3),
               throttles=int(9000 / config.aggregation_concurrency(retry=False)))
def aggregate(event: chalice.app.SQSEvent):
    app.index_controller.aggregate(event)
```

…to be:
```
@app.threshold(errors=int(config.aggregation_concurrency(retry=False) * 3),
               throttles=int(9000 / config.aggregation_concurrency(retry=False)))
@app.on_sqs_message(queue=config.tallies_queue_name(),
                    batch_size=IndexController.document_batch_size)
def aggregate(event: chalice.app.SQSEvent):
    app.index_controller.aggregate(event)
```

…restored the desired thresholds instead of using the `0` default:
```
(.venv) daniel@Crispin ~/repo/azul3 $ egrep 'alarm_name|threshold' terraform/cloudwatch.tf.json
                        ...
                        "alarm_name": "azul-indexer_aggregate_errors-anvildev.alarm",
                        "threshold": 192,
                        ...
                        "alarm_name": "azul-indexer_aggregate_throttles-anvildev.alarm",
                        "threshold": 140,
                        ...
```

--- a/lambdas/indexer/app.py
+++ b/lambdas/indexer/app.py
@@ -246,17 +246,17 @@ def post_notification(catalog: CatalogName, action: str):
     return app.index_controller.handle_notification(catalog, action)
 
 
-@app.on_sqs_message(queue=config.notifications_queue_name(), batch_size=1)
 @app.threshold(errors=int(config.contribution_concurrency(retry=False) * 2 / 3),
                throttles=int(15000 / config.contribution_concurrency(retry=False)))
+@app.on_sqs_message(queue=config.notifications_queue_name(), batch_size=1)
 def contribute(event: chalice.app.SQSEvent):
     app.index_controller.contribute(event)
 
 
-@app.on_sqs_message(queue=config.tallies_queue_name(),
-                    batch_size=IndexController.document_batch_size)
 @app.threshold(errors=int(config.aggregation_concurrency(retry=False) * 3),
                throttles=int(9000 / config.aggregation_concurrency(retry=False)))
+@app.on_sqs_message(queue=config.tallies_queue_name(),
+                    batch_size=IndexController.document_batch_size)
 def aggregate(event: chalice.app.SQSEvent):
     app.index_controller.aggregate(event)
 
@@ -264,10 +264,10 @@ def aggregate(event: chalice.app.SQSEvent):
 # Any messages in the tallies queue that fail being processed will be retried
 # with more RAM in the tallies_retry queue.
 
-@app.on_sqs_message(queue=config.tallies_queue_name(retry=True),
-                    batch_size=IndexController.document_batch_size)
 @app.threshold(errors=int(config.aggregation_concurrency(retry=True) * 1 / 16),
                throttles=0)
+@app.on_sqs_message(queue=config.tallies_queue_name(retry=True),
+                    batch_size=IndexController.document_batch_size)
 def aggregate_retry(event: chalice.app.SQSEvent):
     app.index_controller.aggregate(event, retry=True)
 
@@ -275,10 +275,10 @@ def aggregate_retry(event: chalice.app.SQSEvent):
 # Any messages in the notifications queue that fail being processed will be
 # retried with more RAM and a longer timeout in the notifications_retry queue.
 
-@app.on_sqs_message(queue=config.notifications_queue_name(retry=True),
-                    batch_size=1)
 @app.threshold(errors=int(config.contribution_concurrency(retry=True) * 1 / 4),
                throttles=int(15000 / config.contribution_concurrency(retry=True)))
+@app.on_sqs_message(queue=config.notifications_queue_name(retry=True),
+                    batch_size=1)
 def contribute_retry(event: chalice.app.SQSEvent):
     app.index_controller.contribute(event, retry=True)
 

--- a/src/azul/chalice.py
+++ b/src/azul/chalice.py
@@ -497,6 +497,7 @@ class AzulChaliceApp(Chalice):
 
     def threshold(self, *, errors: int, throttles: int):
         def wrapper(f):
+            assert isinstance(f, chalice.app.EventSourceHandler), f
             f.errors_threshold = errors
             f.throttles_threshold = throttles
             return f

